source/extensions/filters/http/ext_proc/processor_state.h

```diff
@@ -18,12 +18,37 @@ class Filter;
 
 class ProcessorState : public Logger::Loggable<Logger::Id::filter> {
 public:
+  // This describes whether the filter is waiting for a response to a gRPC message
   enum class CallbackState {
-    Idle,
+    Idle = 0,
     // Waiting for a "headers" response
-    Headers,
-    // Waiting for a "body" response
-    BufferedBody,
+    HeadersCallback = 1,
+    // Done processing headers for whatever reason
+    HeadersComplete = 2,
+    // Waiting for a "body" response in buffered mode
+    BufferedBodyCallback = 3,
```

Comment:
Can you remind me if, as with Envoy filter chains, it's possible to have both a headers and body message to the server in-flight at the same time?

Reply:
Not in the ext_proc protocol right now -- the filter goes to some trouble now to wait for the response to the headers message before sending the body message, and to wait for the body before trailers, and to do all that while assuming that the filter might be configured to send any combination of those three. I think that makes things simplest on the server, and it lets the server decide if it wants to see the body at all only after seeing the headers -- I think that's important, actually.

If we end up with a lot of servers that don't want to make this decision dynamically then there are more options that we could add, like pipelining the headers and body messages.